bin/migrate-oats-data/submissions/app_submissions.py

```diff
@@ -0,0 +1,250 @@
+from common import (
+    ALRChangeCode,
+    log_end,
+    log_start,
+)
+from db import inject_conn_pool
+from constants import BATCH_UPLOAD_SIZE
+from psycopg2.extras import execute_batch, RealDictCursor
+import traceback
+from enum import Enum
+
+etl_name = "alcs_app_sub"
+
+@inject_conn_pool
+def process_alcs_app_submissions(conn=None, batch_size=BATCH_UPLOAD_SIZE):
+    """
+    This function is responsible for processing app_submissions in batches, updating records in the alcs.application_submissions table.
+
+    Args:
+    conn (psycopg2.extensions.connection): PostgreSQL database connection. Provided by the decorator.
+    batch_size (int): The number of items to process at once. Defaults to BATCH_UPLOAD_SIZE.
+    """
+
+    log_start(etl_name)
+    with conn.cursor(cursor_factory=RealDictCursor) as cursor:
+        with open(
+            "submissions/sql/app_submission_count.sql",
+            "r",
+            encoding="utf-8",
+        ) as sql_file:
+            count_query = sql_file.read()
+            cursor.execute(count_query)
+            count_total = dict(cursor.fetchone())["count"]
+        print("- Total Application Submission data to insert: ", count_total)
+
+        failed_inserts = 0
+        successful_inserts_count = 0
+        last_application_id = 0
+
+        with open(
+            "submissions/sql/app_submission.sql",
+            "r",
+            encoding="utf-8",
+        ) as sql_file:
+            application_sql = sql_file.read()
+            while True:
+                cursor.execute(
+                    f"{application_sql} WHERE acg.alr_application_id > {last_application_id} ORDER BY acg.alr_application_id;"
+                )
+
+                rows = cursor.fetchmany(batch_size)
+
+                if not rows:
+                    break
+                try:
+                    applications_to_be_inserted_count = len(rows)
+
+                    insert_app_sub_records(conn, batch_size, cursor, rows)
+
+                    successful_inserts_count = (
+                        successful_inserts_count + applications_to_be_inserted_count
+                    )
+                    last_application_id = dict(rows[-1])["alr_application_id"]
+
+                    print(
+                        f"retrieved/inserted items count: {applications_to_be_inserted_count}; total successfully inserted submissions so far {successful_inserts_count}; last inserted application_id: {last_application_id}"
+                    )
+                except Exception as e:
+                    # this is NOT going to be caused by actual data insert failure. This code is only executed when the code error appears or connection to DB is lost
```

Comment:
Actually, it is not true. Unlike the "update", "insert" will return an error if it fails to insert a record due to FK constraint or field type mismatch and etc.